### PR TITLE
Implemented GetAvailableRoutes and BookRide

### PR DESCRIPTION
**Pull request by VeTcRoT**

### CODE REVIEW

File: src/RouteService/RouteService.Persistence/Repositories/RouteRepository.cs [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    public class RouteRepository : BaseRepository<Route>, IRouteRepository
+    {
+        public RouteRepository(RouteServiceDbContext dbContext) : base(dbContext) { }
+
+        public async Task<Route> BookRideAsync(string routeId, string from, string to, int numberOfSeats)
+        {
+            var ride = await _dbContext.Routes
+                .Where(r => r.From == from && r.SeatsAvailable >= numberOfSeats && r.RouteId == routeId)
+                .Join(
+                    _dbContext.Routes.Where(r => r.To == to && r.SeatsAvailable >= numberOfSeats && r.RouteId == routeId),
```
Comment (EugeneEX) [inline src/RouteService/RouteService.Persistence/Repositories/RouteRepository.cs]: This method is more of a business logic than repository

File: src/RouteService/RouteService.Persistence/Repositories/RouteRepository.cs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+            var routeToReturn = new Route()
+            {
+                RouteId = routeId,
+                DepartureTime = ride.RouteFrom.DepartureTime,
+                ArrivalTime = ride.RouteTo.ArrivalTime,
+                From = from,
+                To = to,
+                NumberOfSeats = ride.RouteFrom.NumberOfSeats,
+                SeatsAvailable = ride.RouteFrom.SeatsAvailable + numberOfSeats,
```
Comment (EugeneEX) [inline src/RouteService/RouteService.Persistence/Repositories/RouteRepository.cs]: Why do you need to add number of booked seats to the seats available?

File: src/RouteService/RouteService.Persistence/Repositories/RouteRepository.cs [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+                ExtraInfo = ride.RouteFrom.ExtraInfo + "\n" + ride.RouteTo.ExtraInfo
```
Comment (EugeneEX) [inline src/RouteService/RouteService.Persistence/Repositories/RouteRepository.cs]: Please note that ExtraInfo is the same for the whole route, so using this approach you will have a duplication of ExtraInfo for each section of the route

File: src/RouteService/RouteService.Domain/Entities/Ride.cs [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+{
+    public class Ride
+    {
+        public int Id { get; set; }
+        public string RouteId { get; set; } = string.Empty;
+        public DateTime DepartureTime { get; set; }
+        public DateTime ArrivalTime { get; set; }
+        public string From { get; set; } = string.Empty;
+        public string To { get; set; } = string.Empty;
+        public int NumberOfSeats { get; set; }
```
Comment (EugeneEX) [inline src/RouteService/RouteService.Domain/Entities/Ride.cs:11]: What does NumberOfSeats mean here? The total number of seats on this ride? Or total number of seats on the whole route?

File: src/RouteService/RouteService.Application/Features/Routes/Queries/GetAvailableRoutes/GetAvailableRoutesQueryHandler.cs [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+        public async Task<IEnumerable<Ride>?> Handle(GetAvailableRoutesQuery request, CancellationToken cancellationToken)
+        {
+            var validator = new GetAvailableRoutesQueryValidator();
+            var validationResult = await validator.ValidateAsync(request);
+
+            if (validationResult.Errors.Count > 0)
+                throw new ValidationException(validationResult);
+
+            return await _unitOfWork.RouteRepository.GetAvailableRoutesByQueryAsync(request.From, request.To, request.DepartureTime, request.NumberOfSeats);
```
Comment (EugeneEX) [inline src/RouteService/RouteService.Application/Features/Routes/Queries/GetAvailableRoutes/GetAvailableRoutesQueryHandler.cs]: Please set up a mapping between Ride and RideDto, don't return the entity type right to the controller
In RideDto you can put string ExtraInfo property, not including the whole Route object

File: src/RouteService/RouteService.Domain/Entities/Ride.cs (hunk 4 above)
Comment (VeTcRoT) [reply src/RouteService/RouteService.Domain/Entities/Ride.cs:11]: The total number of seats on this ride

Comment (VeTcRoT) [reply src/RouteService/RouteService.Domain/Entities/Ride.cs:11]: I use this number to produce booked seats

**Review by EugeneEX — APPROVED** [review]

**Pull request merged**